superchain-backend.md

```diff
@@ -0,0 +1,457 @@
+- wip(interop) block dependency graph [`monorepo #10044`](https://github.com/ethereum-optimism/optimism/pull/10044) (April 4)
+    - Stacked on top of `#9612`.
+    - `BlockDependencies` that holds on to safety information.
+    - Some doubts about the `dependents` map with key uniqueness / reorg safety.
+    - Global lock on data.
+    - Recursive block-safety algorithm, unable to handle loops.
+    - Centered around block-safety.
+    - Needs work for determining safety of initiating messages transitively efficiently.
+    - All in-memory, no back-pressure or limits.
+
+From the above we can use the draft of the standalone backend service, but will need to change:
+- The executing message parsing: detect from receipts instead
+- The safety-resolution: use an append-log model instead, to avoid recursion
+- Maintain only recent history in memory, flush the older history to disk, to not go OOM.
+- Revisit reorg protection: we do need checkpoints to determine how to roll back the log,
+  but should be able to do so without the current separate data mappings.
+
+# Proposed Solution
+
+<!-- *A high level overview of the proposed solution.
+When there are multiple alternatives there should be an explanation of why one solution was picked over other solutions.
+As a rule of thumb, including code snippets (except for defining an external API) is likely too low level.* -->
+
+## Closing draft PRs
+
+We close the specs PR and the 3 draft PRs.
+Parts of the implementation are still applicable, this can be cherry-picked with attribution.
+
+## Superchain Backend specs
+
+A new specs PR should describe the interface (without enshrining batching),
+as well as a simplified version of a standalone backend.
+
+Also note that the specs should be updated to cover the edge-cases of intra-block event ordering.
+An invariant for event-log index should be introduced.
+
+The interface:
+```
+type SafetyLevel # see design-doc
+
+type Verifier interface {
+    CheckMessage(Identifier, PayloadHash) SafetyLevel
+    CheckBlock(ChainID, BlockHash, BlockNumber) SafetyLevel
+}
+```
+
+Summary of simplified implementation to include in specs:
+- One append-log per chain
+- Safety labels (pointer to block-number) per append-log
+- Hashing method to turn event-log (initiating or executing message) into versioned hash
+- Append-log entries as follows:
+  - Block metadata (prefixed to be distinguishable from event-log hashes)
+  - Event log hashes (concatenated)
+- Binary search to find a log entry by block-number
+- Reorgs by truncating the append-log to the last common block before the reorg divergence
+
+## Superchain Backend implementation: `op-resolver`
+
+> *Name suggestions welcome, `op-superchain` seems to generic however.*
+
+This service resolves superchain dependencies, and implements the above proposed spec for a standalone implementation.
+
+A later iteration of this can support the ideated shreds streaming,
```

Comment:
Perhaps in the future this could also support the p2p optimizations, ie participate in the p2p network and fetch data that way